[4.0] Feature/New Publishing Workflow

bembelimen opened the pull request:
Hello everyone,

with the Google Summer of Code 2017 project, we implemented a new workflow extension in Joomla! 4, which replaces the current publish/unpublish/trashed/archived state. We want to implement a flexible and unlimited workflow which can be extended for other extensions, too.

Presentation at the JWC 2017: https://www.youtube.com/watch?v=5e-HECq44B0
Presentation at JAB 2018: https://youtu.be/z-TfSEggUZM

### Additional Information
https://docs.joomla.org/Publishing_Workflow
https://docs.joomla.org/GSOC_2017_Project_Ideas#Project_VI:_Publishing_Workflow_in_Joomla.21

### Testing Instructions
Install the PR via the Patchtester and reinstall your Joomla! or download the ZIP file (https://github.com/bembelimen/joomla-cms/archive/publishing-workflow.zip) and make a fresh installation. Afterwards go to the Article manager, there you'll find a new menu entry "Workflows", play with it around. You can also use the implemented default workflow.

Workflows will be assigned in the category parameters.

### Images

![grafik](https://user-images.githubusercontent.com/1229869/39334713-0c89d844-49b0-11e8-8198-f6e3334acb3b.png)

![grafik](https://user-images.githubusercontent.com/1229869/39334741-29ef97c0-49b0-11e8-9d5a-6b955d390e9b.png)

![grafik](https://user-images.githubusercontent.com/1229869/39334758-46ca432c-49b0-11e8-94ac-07840067a7a6.png)

Comments, tips, support would be very appreciated!

Buddhima committed:
fixing current date function issue

bembelimen committed:
Merge pull request #105 from Buddhima/fix_create_time_sql

Fix create time sql

bembelimen committed:
Fix context when saving a workflow

bembelimen committed:
Merge pull request #106 from joomla-projects/Fix_context_while_saving

Fix context when saving a workflow

bembelimen committed:
Create a default state while creating a new workflow

HRIT-Florian committed:
save2copy button added
CanDo save2copy
Code rerranged

bembelimen committed:
Rename "workflow" to "workflows" in the menu

HRIT-Florian committed:
Merge branch 'staging' of github.com:joomla-projects/GSoC17_publishing_workflow into save_copy

HRIT-Florian committed:
Auto-populate the model state. filter.extension

bembelimen committed:
Fix context for state model

HRIT-Florian committed:
Merge branch 'staging' of github.com:joomla-projects/GSoC17_publishing_workflow into save_copy

HRIT-Florian committed:
save2copy workfolow > states

bembelimen committed:
Fix saving context in transiton model

HRIT-Florian committed:
Merge branch 'staging' of github.com:joomla-projects/GSoC17_publishing_workflow into save_copy

bembelimen committed:
Add number of transitions/states in workflows view

HRIT-Florian committed:
Merge branch 'staging' of github.com:joomla-projects/GSoC17_publishing_workflow into save_copy

HRIT-Florian committed:
Clean up code

HRIT-Florian committed:
save2copy workflow > transitions

HRIT-Florian committed:
generateNewTitle when save2copy added

bembelimen committed:
Merge pull request #107 from Didldu-Florian/save_copy

Save2copy

bembelimen committed:
Add published filter for status/transition listing in workflow view

bembelimen committed:
Fix trashed status when filtered

bembelimen committed:
Merge transition and state column with special transition

chmst committed:
Use plurals for the sidebar

bembelimen committed:
Align states to the left

bembelimen committed:
Converted articles state filter to a grouped list

bembelimen committed:
Merge pull request #109 from joomla-projects/transitions_in_article_list

Merge transition and state column in articles view

bembelimen committed:
Sniffer satisfaction

bembelimen committed:
Sniffer satisfaction

bembelimen committed:
Code cleanup

bembelimen committed:
Fix articles count in category

bembelimen committed:
Merge pull request #115 from joomla-projects/fix_sniffer

Sniffer satisfaction

Commit:
Fix execute transitions on all articles #112 (#117)

* Fix execute transitions on all articles #112

* Fix minor issue

bembelimen committed:
Rename extenstion => extension

chmst committed:
Improve state edit layout - move condition from publishing area to main area

chmst committed:
Label with generic text

Buddhima committed:
adding new workflow states field

Buddhima committed:
fix codesniffer issues

bembelimen committed:
Merge branch 'staging' of github.com:joomla-projects/GSoC17_publishing_workflow into grouped_state_filter

bembelimen committed:
Merge pull request #116 from joomla-projects/fix_articles_count_in_category

Fix articles count in category

bembelimen committed:
Merge pull request #118 from chmst/Move_field_condition

Move field condition, improve UX

Buddhima committed:
update field class according to suggesstions

bembelimen committed:
Merge pull request #119 from Buddhima/grouped_state_filter_field

Adding new workflow states field

bembelimen committed:
Merge pull request #111 from joomla-projects/grouped_state_filter

Converted articles state filter to a grouped list

Buddhima committed:
fix issue with settting the default value with state edit form

bembelimen committed:
Merge pull request #120 from Buddhima/fix_issue_states_default

Fix issue with settting the default value with state edit form

chmst committed:
Reolve https://github.com/joomla-projects/GSoC17_publishing_workflow/issues/114

chmst committed:
Code changed: getTable, use sprintf

HRIT-Florian committed:
Unpublish new copy

bembelimen committed:
Merge pull request #123 from Didldu-Florian/save_copy

Save2Copy

HRIT-Florian committed:
Save2copy for states

HRIT-Florian committed:
save2copy check if title exists

bembelimen committed:
Merge pull request #122 from chmst/more_plurals_in_sidebar

Use plurals for the sidebar

chmst committed:
Disable the switcher on edit state if the state is default

chmst committed:
Fixed getValue

bembelimen committed:
Merge pull request #125 from chmst/disable_default-switcher_on_edit_state

Disable the switcher on edit state if the state is default

chmst committed:
Simplified and escaped

chmst committed:
fixed: check for table-entry workflow

bembelimen committed:
Merge pull request #121 from chmst/Workflow_name_in_lists

Resolve issue 114

zero-24 committed:
merge 4.0-dev

zero-24 committed:
recompile core.min.js

bembelimen committed:
Fix new class structure

bembelimen committed:
Fix #110 deletion of trashed articles

bembelimen committed:
Readd featuring articles

Commit:
Adds ordering to the states and transitions #94

Buddhima committed:
adding current state to article edit view, transition drop-down

Buddhima committed:
remove default value in XML config

bembelimen committed:
Merge pull request #128 from Buddhima/improve_article_edit_transition

Adding current state to article edit view, transition drop-down

bembelimen committed:
Shorten return code

bembelimen committed:
Merge pull request #129 from joomla-projects/improve_article_edit_transition

Improve article edit transition

Commit:
Merge remote-tracking branch 'origin/staging' into ordering

Buddhima committed:
adding workflow_associations table structure

bembelimen committed:
Moved form xml to "forms" folder

bembelimen committed:
Moved validation to the table class
Load workflow_id only from the session when not set (while saving)
Improve disabled default state

bembelimen committed:
Merge branch '4.0-dev' of github.com:joomla/joomla-cms into staging

# Conflicts:
#	components/com_content/Model/ArchiveModel.php
#	components/com_content/Model/ArticlesModel.php
#	media/system/js/core.min.js

bembelimen committed:
Add deploy version instead of a normal version

bembelimen committed:
Add more deploy versions

bembelimen committed:
Merge pull request #124 from Didldu-Florian/save_copy

Save copy

bembelimen committed:
Merge branch 'staging' into improve_saving

# Conflicts:
#	administrator/components/com_workflow/Model/State.php

Buddhima committed:
fix sql issue

Buddhima committed:
adding new workflow-association manipulations

Buddhima committed:
change article saving to use associations

bembelimen committed:
Fix saving with workflow_id

bembelimen committed:
Merge pull request #134 from joomla-projects/Deploy_version

Add deploy version instead of a normal version

bembelimen committed:
Merge pull request #133 from joomla-projects/improve_saving

Moved validation to the table class

Buddhima committed:
changing article saving, transition saving actions to do via associations table

Buddhima committed:
update state filtering field to use workflow_associations

Buddhima committed:
fix issue with identifying transitionId when executing runTransition

bembelimen committed:
First rewrite to reflect the new class convention

bembelimen committed:
Change table class where needed

bembelimen committed:
Fix wrong language string

bembelimen committed:
More table fixes

bembelimen committed:
Next table classname fix

bembelimen committed:
Table::getInstance => new Table

bembelimen committed:
Sniffer fixes

bembelimen committed:
Fix where in workflow table

bembelimen committed:
Fix where in workflow table

bembelimen committed:
Fix saving error for workflows

Buddhima committed:
adding comments and remove an unnecessary method

chmst committed:
Move the main properties auf a transition from the right column to the main panel.

Buddhima committed:
Merge pull request #135 from Buddhima/workflow_associations_impl

Workflow associations implementation

Buddhima committed:
move namespace to state filter field

Buddhima committed:
update featured article view according to articles view changes

Buddhima committed:
fixing style issue in featured default

Buddhima committed:
style fixing in featured default 2

Buddhima committed:
style fixing in featured default 3

Commit:
Merge branch 'staging' of https://github.com/joomla-projects/GSoC17_publishing_workflow into ordering

# Conflicts:
#	administrator/components/com_workflow/Field/TransitionField.php

bembelimen committed:
Merge branch 'staging' into ordering

# Conflicts:
#	administrator/components/com_workflow/Field/TransitionField.php

bembelimen committed:
Set default values

bembelimen committed:
Merge branch 'staging' into ordering

bembelimen committed:
Move table for workflow association

bembelimen committed:
Fix sql default value

bembelimen committed:
Move table for workflow association

bembelimen committed:
Merge branch 'staging' into ordering

# Conflicts:
#	installation/sql/mysql/joomla.sql

bembelimen committed:
Merge branch '4.0-dev' of github.com:joomla/joomla-cms into staging

bembelimen committed:
Remove colored workflow param select

bembelimen committed:
Merge branch 'staging' of github.com:joomla-projects/GSoC17_publishing_workflow into ordering

bembelimen committed:
Add workflow ordering

bembelimen committed:
Merge pull request #136 from chmst/improve-transition-edit-view

Move the main properties auf a transition from the right column to the main panel

Buddhima committed:
Merge pull request #137 from Buddhima/update_featured_article_view_with_associations

Update featured article view with associations

Buddhima committed:
adding final state in transition drop-down

Buddhima committed:
adding workflow helper to return condition name and append it to the transition text

Commit:
Add ordering to the state and transition

Commit:
Merge branch 'ordering' of https://github.com/joomla-projects/GSoC17_publishing_workflow into ordering

bembelimen committed:
Add condition to the transition field
Remove double JText

Buddhima committed:
Merge pull request #1 from joomla-projects/add_state_in_transition_text

Add condition to the transition field

Buddhima committed:
Merge pull request #142 from Buddhima/add_state_in_transition_text

Add condition in transition text

Buddhima committed:
implementing delete record in workflow associations table when deleting articles

Buddhima committed:
Merge pull request #144 from Buddhima/impl_workflow_associations_delete

Implement workflow associations delete when deleting articles

Buddhima committed:
update admin mod_latest to use workflow_state

Commit:
Fixing issues from review

Commit:
Merge remote-tracking branch 'origin/staging' into ordering

# Conflicts:
#	administrator/components/com_workflow/Field/TransitionField.php

Commit:
Fixing conlict with staging

Commit:
Fixing projection to int on workflow_id

bembelimen committed:
Merge pull request #143 from joomla-projects/ordering

Ordering

Buddhima committed:
fixing issues with using workflow_associations in category view

Buddhima committed:
update archive articles modulw with workflow_association changes

Buddhima committed:
fixing transition field ambiguity in ordering

Buddhima committed:
Merge pull request #149 from Buddhima/fix_transition-field_ordering

Fixing transition field ambiguity in ordering

Buddhima committed:
Merge pull request #147 from Buddhima/fixing_category_list_view

fixing issues with using workflow_associations in category view

Buddhima committed:
adding layouts for batch processing

Buddhima committed:
fixing display issues with workflow states dropdown

Buddhima committed:
adding workfow state change functionality

Buddhima committed:
adding access control checks for workflows list related places

Buddhima committed:
few modifications to workflow model

Buddhima committed:
update controller and models of state, transition

Buddhima committed:
modify config files according to com_workflow permission hierarchy

Buddhima committed:
refactor models and views to use permissions

Buddhima committed:
fix suggesstion in PR

Buddhima committed:
fix for adding users

Buddhima committed:
adding transition.run permission to com_content entry

Buddhima committed:
fixing workflow saving issue and use permissions for list view state changes

Buddhima committed:
fix suggesstion in PR-2

Buddhima committed:
fixing suggesstion in PR

bembelimen committed:
Merge branch 'master' into staging

# Conflicts:
#	administrator/components/com_categories/tmpl/categories/default.php
#	administrator/components/com_content/Model/ArticlesModel.php
#	administrator/components/com_content/tmpl/articles/default.php
#	components/com_content/Model/ArchiveModel.php
#	components/com_content/Model/ArticlesModel.php
#	components/com_content/tmpl/category/blog.php
#	components/com_content/tmpl/category/blog_item.php
#	components/com_content/tmpl/featured/default.php
#	installation/sql/mysql/joomla.sql
#	media/system/js/core.min.js

bembelimen committed:
Merge branch '4.0-dev' into 4.0_alpha_integration

# Conflicts:
#	administrator/components/com_content/Model/ArticlesModel.php
#	administrator/components/com_content/tmpl/articles/default.php
#	administrator/index.php
#	administrator/modules/mod_menu/tmpl/default.php
#	administrator/templates/atum/js/template.js
#	components/com_content/Model/ArchiveModel.php
#	components/com_content/Model/ArticlesModel.php
#	components/com_content/forms/filter_articles.xml
#	components/com_content/tmpl/category/blog_item.php
#	grunt-settings.yaml
#	index.php
#	installation/index.php
#	installation/sql/mysql/joomla.sql
#	media/system/js/core.min.js
#	media/vendor/tinymce/changelog.txt
#	media/vendor/tinymce/plugins/help/plugin.min.js
#	media/vendor/tinymce/plugins/image/plugin.js
#	media/vendor/tinymce/plugins/image/plugin.min.js
#	media/vendor/tinymce/plugins/imagetools/plugin.js
#	media/vendor/tinymce/plugins/imagetools/plugin.min.js
#	media/vendor/tinymce/plugins/nonbreaking/plugin.js
#	media/vendor/tinymce/plugins/nonbreaking/plugin.min.js
#	media/vendor/tinymce/plugins/paste/plugin.js
#	media/vendor/tinymce/plugins/paste/plugin.min.js
#	media/vendor/tinymce/plugins/table/plugin.js
#	media/vendor/tinymce/plugins/table/plugin.min.js
#	media/vendor/tinymce/plugins/template/plugin.js
#	media/vendor/tinymce/plugins/template/plugin.min.js
#	media/vendor/tinymce/plugins/visualchars/plugin.js
#	media/vendor/tinymce/plugins/visualchars/plugin.min.js
#	media/vendor/tinymce/themes/inlite/theme.js
#	media/vendor/tinymce/themes/inlite/theme.min.js
#	media/vendor/tinymce/themes/mobile/theme.js
#	media/vendor/tinymce/themes/mobile/theme.min.js
#	media/vendor/tinymce/themes/modern/theme.js
#	media/vendor/tinymce/themes/modern/theme.min.js
#	media/vendor/tinymce/tinymce.js
#	media/vendor/tinymce/tinymce.min.js
#	package.json
#	plugins/editors/tinymce/tinymce.xml
#	templates/cassiopeia/css/template.css
#	templates/cassiopeia/js/template.js

bembelimen committed:
Merge pull request #155 from joomla-projects/4.0_alpha_integration

4.0 alpha integration

bembelimen committed:
Rename transition.run => core.execute.transition
Move transition action language string to the Joomla! core language file

bembelimen committed:
Merge pull request #157 from joomla-projects/impl_access_control

Impl access control

bembelimen committed:
Merge branch 'staging' into update_archive_article_with_associations

bembelimen committed:
Add parameter to WorkflowState field, so it's possible to show all states or only the active one
Add extension filter parameter to WorflowState field
Remove inline-form class to archive view
Renamed archived module

bembelimen committed:
Merge pull request #158 from joomla-projects/update_archive_article_with_associations

Update archive article with associations

bembelimen committed:
Merge branch 'staging' into implementing_batch_processing

bembelimen committed:
Add permission check for batch transition executions

bembelimen committed:
Fix broken transition batch

bembelimen committed:
Merge pull request #159 from joomla-projects/implementing_batch_processing

Implementing batch processing

bembelimen committed:
Merge pull request #145 from Buddhima/update_admin_module

update admin mod_latest to use workflow_state

bembelimen committed:
Set transition batch execution to super users only

bembelimen committed:
Implement some cleanups

bembelimen committed:
Fix article saving

bembelimen committed:
Implement email notification for transition execution

bembelimen committed:
Merge pull request #160 from joomla-projects/email_notification

Implement email notification for transition execution

bembelimen committed:
Rename the transition dropdown in the article edit form

bembelimen committed:
Readd save2copy

bembelimen committed:
Convert frontend state to transition in edit form

bembelimen committed:
Merge branch '4.0-dev' into staging

# Conflicts:
#	administrator/language/en-GB/en-GB.com_content.ini
#	media/system/js/core.min.js
#	modules/mod_articles_archive/Helper/ArticlesArchiveHelper.php

bembelimen committed:
Merge 4.0 dev from core

bembelimen committed:
Merge branch '4.0-dev' into staging

# Conflicts:
#	administrator/components/com_content/access.xml
#	administrator/components/com_menus/presets/joomla.xml
#	administrator/components/com_menus/presets/modern.xml
#	administrator/language/en-GB/en-GB.com_content.ini
#	administrator/language/en-GB/en-GB.ini
#	components/com_content/Model/ArchiveModel.php
#	media/system/js/core.min.js

bembelimen committed:
Merge branch 'staging' of github.com:joomla-projects/GSoC17_publishing_workflow into publishing-workflow

bembelimen committed:
Fix sniffer issues

bembelimen committed:
Remove unnecessary parenthesis

bembelimen committed:
Fix more sniffer issues

bembelimen committed:
Merge branch 'publishing-workflow' of github.com:bembelimen/joomla-cms into publishing-workflow

bembelimen committed:
Rename setHome to setDefault

bembelimen committed:
Use JDate instead of date()

bembelimen committed:
Protect workflow from uninstall

bembelimen committed:
Add ordering to the select for workflow state field

bembelimen committed:
Fix menu  with deactivated workflow

chmst committed:
Replace $db->qn by $db->quoteName

chmst committed:
Replace $db->q( by $db->quote(

bembelimen committed:
Merge pull request #6 from chmst/qn-to-quoteName

Improve Code - use quoteName and quote 

Thank you @chmst !

bembelimen committed:
Validate before status deletion if no items are assigned
Validate before status change to trashed if no items are assigned

bembelimen committed:
Moved transition form field to the libraries folder

bembelimen committed:
Moved transition form field to the libraries folder

bembelimen committed:
Implement Workflow library class with condition constants
Fix minor html errors

bembelimen committed:
Merge branch 'publishing-workflow' into publishing-workflow-library

bembelimen committed:
Merge pull request #10 from bembelimen/publishing-workflow-library

Publishing workflow library

bembelimen committed:
Merge branch '4.0-dev' of github.com:joomla/joomla-cms into publishing-workflow

# Conflicts:
#	administrator/components/com_content/Helper/ContentHelper.php
#	administrator/components/com_content/View/Articles/HtmlView.php

bembelimen committed:
Fix workflowHelper Submenu call

bembelimen committed:
Add Workflow constant to the frontpage

bembelimen committed:
Moves state select field to the library folder

bembelimen committed:
Fix some sniffer errors

bembelimen committed:
Fix wrong parameter structure

bembelimen committed:
Merge branch '4.0-dev' of github.com:joomla/joomla-cms into publishing-workflow

bembelimen committed:
Fix wrong input parameter

bembelimen committed:
Fix transition filter in the form

wilsonge committed:
Fix state transitions on publishing content

bembelimen committed:
Merge pull request #11 from wilsonge/fix/publish-transitions

Fix state transitions on publishing content

bembelimen committed:
Add "All" for source states in transitions

bembelimen committed:
Fix wrong table prefix

bembelimen committed:
Fix frontend article displaying

brianteeman commented on administrator/language/en-GB/en-GB.com_content.ini:
```diff
@@ -3,15 +3,17 @@
 COM_CONTENT="Articles"
-COM_CONTENT_ARTICLE_CONTENT="Content"
 COM_CONTENT_ARTICLES_TITLE="Articles"
+COM_CONTENT_ARTICLE_CONTENT="Content"
```
please revert this change

brianteeman commented on administrator/language/en-GB/en-GB.com_content.ini:
```diff
@@ -3,15 +3,17 @@
 COM_CONTENT_BATCH_TIP="If a category is selected for move/copy, any actions selected will be applied to the copied or moved articles. Otherwise, all actions are applied to the selected articles."
 COM_CONTENT_CHANGE_ARTICLE="Select or Change article"
 COM_CONTENT_CHANGE_ARTICLE_BUTTON="Select/Change"
+COM_CONTENT_CONFIGURATION="Articles: Options"
```
please revert this change

brianteeman commented on administrator/language/en-GB/en-GB.com_content.ini:
```diff
@@ -20,38 +22,46 @@ COM_CONTENT_CONFIG_CATEGORY_SETTINGS_DESC="These settings apply for Articles Cat
+COM_CONTENT_EDITORCONFIG_FIELDSET_LABEL="Configure Edit Screen"
+COM_CONTENT_EDIT_ARTICLE="Edit Article"
 COM_CONTENT_ERROR_ALL_LANGUAGE_ASSOCIATED="A content item set to All languages can't be associated. Associations have not been set."
+COM_CONTENT_ERROR_UPDATE_STATE="Can not set state to the item"
```
==> Can't set the state of the item

brianteeman commented:
I will do a PR against your branch next week to make the language files correct english.

Can you please change the qn to quotename etc

bembelimen committed:
Fix broken status/transition filtering

bembelimen committed:
Fix transition publishing/ordering

bembelimen committed:
Change qn => quoteName

bembelimen committed:
Change qn => quoteName

bembelimen committed:
Add inheritation for workflows in categories

bembelimen committed:
Create Joomla! 3 => 4 migration script for com_workflow

bembelimen committed:
Add extension entry for update proccess

bembelimen commented on administrator/components/com_admin/sql/updates/mysql/4.0.0-2018-05-15.sql:
```diff
@@ -0,0 +1,117 @@
+-- Creating Associations for existing content
+--
+INSERT INTO `#__workflow_associations` (`item_id`, `state_id`, `extension`)
+SELECT `id`, CASE WHEN `state` = -2 THEN 3 WHEN `state` = 0 THEN 1 WHEN `state` = 2 THEN 4 ELSE 2 END, 'com_content' FROM `#__content`;
```
Hello @nikosdion as discussed on JAB in Cologne, I added the migration INSERT. Could you please look over the file and see, if there is something to improve? Perhaps we have to handle the asset_id in a seperate way? Thank you very much!

nikosdion replied:
Hi Benjamin! Sorry for the late reply. I got a gastrointestinal infection right after JaB and I'm still quite awful.

This query is exactly what I had in mind when we talked! :+1:

bembelimen replied:
Thank you!

bembelimen committed:
Fix error in featured article view

bembelimen committed:
Add missing namespace

carlitorweb commented:
@bembelimen can you fix the conflict files here, for test it?

brianteeman commented:
I tried to test this and on apply the database changes I got

"1101 BLOB, TEXT, GEOMETRY or JSON column 'description' can't have a default value"

brianteeman commented:
Having two columns with states (states and 0)  is confusing and probably not needed

<img width="848" alt="chrome_2018-05-24_09-01-43" src="https://user-images.githubusercontent.com/1296369/40472369-30d6a20e-5f31-11e8-8582-2e63caef6c06.png">

I would remove the second one (0) and at the same time give the transitions column a name and not just an icon

brianteeman commented:
The workflows, transitions and states should have "Names" not "Titles"

COM_WORKFLOW_TITLE="Title"
==>
COM_WORKFLOW_TITLE="Name"

brianteeman commented:
On a new item there is no visible state - this is confusing and unexpected. For me a new article should have a state displayed and this should be the default for that user

<img width="451" alt="chrome_2018-05-24_09-10-32" src="https://user-images.githubusercontent.com/1296369/40472823-6e89d304-5f32-11e8-9a41-d90b0abb348b.png">

<img width="503" alt="chrome_2018-05-24_09-07-55" src="https://user-images.githubusercontent.com/1296369/40472713-29514308-5f32-11e8-9ec1-dfff50389dda.png">

brianteeman commented:
It is confusing that sometimes you write State and sometimes you wrote Status

brianteeman commented:
I dont find using the state icon to open the select list to be intuitive. Maybe its just me. Maybe it should either not display the text until the select is displayed OR it should always display the select. A better UX than me can advise ;)

![toggle](https://user-images.githubusercontent.com/1296369/40473009-f4cecc6c-5f32-11e8-88de-88047a5a9f23.gif)

brianteeman commented:
Has this been abandoned?

bembelimen commented:
No? There will be some updates this months.

Benjamin Trenkle
Lackerbauerstr. 2
81241 München

Tel: 089/82074746
E-Mail: kontakt@benjamintrenkle.de
Homepage: www.benjamintrenkle.de

Am 01.07.2018 um 22:43 schrieb Brian Teeman:
>
> Has this been abandoned?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/joomla/joomla-cms/pull/20381#issuecomment-401631909>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/ABLELTNfbByiVCsP3uXYb_mdXqpwP4ccks5uCTRygaJpZM4T8s3n>.
>

priiish committed:
Updated files 1.0

priiish committed:
Finished updating language files.

bembelimen committed:
Merge pull request #12 from priiish/publishing-workflow

Updated language file

brianteeman commented:
It would be great if the merge conflicts could be resolved so that we can test the latest changes

bembelimen commented:
We're working on it. We have to rebuild the implementation of com_content

priiish committed:
Language File Update

bembelimen committed:
Merge pull request #14 from priiish/publishing-workflow

Language File Update

priiish committed:
Language File Update

priiish committed:
when trying to delete default state, false error message thrown
-> fixed error message (must be COM_WORKFLOW_MSG_DELETE_DEFAULT)
-> fixed conformation message. now only if number of items deleted is
greater than 0 a message will be displayed => otherwise confusing for user

bembelimen committed:
Merge branch '4.0-dev' into publishing-workflow

# Conflicts:
#	administrator/components/com_categories/Model/CategoryModel.php
#	administrator/components/com_content/Extension/ContentComponent.php
#	administrator/components/com_content/Helper/ContentHelper.php
#	administrator/components/com_content/Model/ArticleModel.php
#	administrator/components/com_content/Model/ArticlesModel.php
#	administrator/components/com_content/View/Articles/HtmlView.php
#	administrator/components/com_content/View/Featured/HtmlView.php
#	administrator/components/com_content/tmpl/articles/default.php
#	administrator/components/com_content/tmpl/articles/default_batch_body.php
#	administrator/components/com_content/tmpl/articles/default_batch_footer.php
#	administrator/components/com_content/tmpl/articles/modal.php
#	administrator/components/com_content/tmpl/featured/default.php
#	administrator/components/com_menus/presets/joomla.xml
#	administrator/components/com_menus/presets/modern.xml
#	administrator/components/com_messages/Model/MessageModel.php
#	administrator/language/en-GB/en-GB.com_content.ini
#	administrator/language/en-GB/en-GB.mod_menu.ini
#	administrator/modules/mod_latest/tmpl/default.php
#	administrator/modules/mod_menu/Menu/CssMenu.php
#	components/com_content/Model/ArticlesModel.php
#	components/com_content/Model/FeaturedModel.php
#	components/com_content/tmpl/archive/default.php
#	components/com_content/tmpl/article/default.php
#	components/com_content/tmpl/category/blog_item.php
#	components/com_content/tmpl/category/default_articles.php
#	installation/sql/mysql/joomla.sql
#	language/en-GB/en-GB.mod_articles_archive.ini
#	language/en-GB/en-GB.mod_articles_archive.sys.ini
#	libraries/src/Categories/CategoriesServiceTrait.php
#	media/system/js/core.min.js
#	modules/mod_articles_archive/Helper/ArticlesArchiveHelper.php

chmst committed:
Fix sniffer errors

chmst committed:
Fix param type

bembelimen committed:
Merge pull request #18 from chmst/workflow-satisfy-sniffer

Fix sniffer errors

bembelimen committed:
Satisfy sniffer

bembelimen committed:
Satisfy sniffer

bembelimen committed:
Satisfy sniffer

chmst committed:
Added the workflow to the main menu because due to some changes in the v4.0 menu there is no longer a sidebar for com_content.

bembelimen committed:
Merge branch '4.0-dev' of github.com:joomla/joomla-cms into publishing-workflow

chmst committed:
Remove duplicate Lines in use statements

bembelimen committed:
Merge pull request #20 from chmst/duplicate-lines-modelcategory

Remove duplicate Lines in use statements

bembelimen committed:
Merge pull request #15 from priiish/publishing-workflow

Language File Update

bembelimen committed:
Merge pull request #16 from priiish/publishing_workflow_bugfix

Fixed error messages

bembelimen committed:
Merge pull request #19 from chmst/add-workflow-to-predefined-xml

Added the workflow to the main menu because due to some changes in th…

bembelimen committed:
Fix loading ContentHelper twice

bembelimen committed:
Merge branch '4.0-dev' of github.com:joomla/joomla-cms into publishing-workflow

# Conflicts:
#	components/com_content/tmpl/archive/default.php

bembelimen commented:
@brianteeman we have here two different things with the same wording:
1. The "Status" which representate the status of the current viewed item (like a banner is published, a workflow itself is unpublished etc.)
2. The "States" which are the new "layer" in the workflow, where you define your workflow states

I would like to have the 2nd one completly renamed into something "unique" which don't leads to confusion ("All states have a status" e.g.)

Is there any good English term which describes the functionality of "States" without confusion with "Status"?

bembelimen committed:
Fix broken use calls

bembelimen committed:
Renamed "Title" => "Name"

brianteeman commented:
I agree 100% with your observation about the similarity and confusion of using status and states although they are technically the correct terms

I will take a look at what others do in this case

rdeutz committed:
Merge branch '4.0-dev' into publishing-workflow

chmst committed:
Improve accessibility of list Views. sope=cols, no table footer, pagination not in table

bembelimen committed:
Merge branch '4.0-dev' of github.com:joomla/joomla-cms into publishing-workflow

# Conflicts:
#	build/media/system/js/core.js

bembelimen committed:
Merge pull request #21 from chmst/a11y-list-view-tables

Improve accessibility of list Views. sope=cols, no table footer, pagi…

brianteeman commented:
I am still getting my head around it but
> The "States" which are the new "layer" in the workflow, where you define your workflow states

Could that be stages?
eg
Stage 1 create content
Stage 2 review content
Stage 3 Publish content

bembelimen committed:
Fix layout

wilsonge commented:
I'm going to merge this to push this forward as there are issues keeping this in sync. There's clearly several issues still to work out but people can create PRs as they go

brianteeman commented:
For the record I dont agree with this being merged in its current state and would like to see it reverted and instead of being on a personal repo moved to a joomla-projects repo like all the others and then issues can be worked on there.

This is nowhere near mergeable

infograf768 commented:
This PR has totally broken a new clean 4.0 installation when trying to create articles.
`500 'v40_multi.wa.state_id' isn't in GROUP BY`
```
0 Argument 1 passed to Joomla\Utilities\ArrayHelpergetColumn() must be of the type array, boolean given, called in /Applications/MAMP/htdocs/installmulti/joomla40/administrator/components/com_content/Model/ArticlesModel.php on line 483
```

alikon commented:
+ no postgresql

franz-wohlkoenig commented:
@infograf768 similar #21474

infograf768 commented:
@franz-wohlkoenig 

As I have done a fully clean install from 4.0-dev, the new tables exist here.
Issue is in code.

franz-wohlkoenig commented:
a nightly didn't work so a clean Install is better?

infograf768 commented:
My issue above is when using multilang.
Creating an issue now

infograf768 commented:
See https://github.com/joomla/joomla-cms/issues/21475

brianteeman commented:
Sorry but this is way beyond not ready it is not correct #21476

alikon commented on administrator/components/com_admin/sql/updates/mysql/4.0.0-2018-05-15.sql:
```diff
@@ -0,0 +1,124 @@
+  `asset_id` int(10) DEFAULT 0,
+  `published` tinyint(1) NOT NULL DEFAULT 0,
+  `title` varchar(255) NOT NULL,
+  `description` text NOT NULL DEFAULT '',
```
The MySQL 'TEXT' type cannot have a DEFAULT value (other than NULL)

alikon commented on administrator/components/com_admin/sql/updates/mysql/4.0.0-2018-05-15.sql:
```diff
@@ -0,0 +1,124 @@
+  KEY `created_by` (`created_by`),
+  KEY `modified` (`modified`),
+  KEY `modified_by` (`modified_by`)
+) ENGINE=InnoDB COLLATE=utf8mb4_unicode_ci;
```
shouldn't be `ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 DEFAULT COLLATE=utf8mb4_unicode_ci` ?

laoneo commented:
Please open new issues or pull requests regarding the new workflow feature. This pr got merged and all comments will likely get forgotten.

alikon commented:
ok will do, but ......

roland-d commented on layouts/joomla/edit/global.php:
```diff
@@ -40,6 +40,11 @@
 	'version_note',
 );
 
+if ($displayData->get('item')->id !== null)
```
@bembelimen Why is this change here? This causes a notice undefined in the plugins when you edit them:
<img width="449" alt="image" src="https://user-images.githubusercontent.com/359377/43989148-9c18d22a-9d44-11e8-9540-ffde0d0ad8bf.png">

brianteeman commented:
Please open new issues or pull requests regarding the new workflow feature. This pr got merged and all comments will likely get forgotten.
